crates/matrix-sdk-ui/src/timeline/inner.rs

```diff
@@ -602,21 +605,22 @@ impl<P: RoomDataProvider> TimelineInner<P> {
 
         let is_error = matches!(send_state, EventSendState::SendingFailed { .. });
 
-        let new_item = TimelineItem::Event(item.with_kind(local_item.with_send_state(send_state)));
-        state.items.set(idx, Arc::new(new_item));
+        let new_item = item.with_inner_kind(local_item.with_send_state(send_state));
+        state.items.set(idx, new_item);
 
         if is_error {
             // When there is an error, sending further messages is paused. This
             // should be reflected in the timeline, so we set all other pending
             // events to cancelled.
             let num_items = state.items.len();
             for idx in 0..num_items {
-                let Some(item) = state.items[idx].as_event() else { continue };
-                let Some(local_item) = item.as_local() else { continue };
+                let item = state.items[idx].clone();
```

Comment:
could this `for item in state.items.iter().cloned() {`? It could avoid a dummy bounds check at line 617.

Reply:
No, the iterator will borrow `state.items`. However, we could do `for item in (*state.items).clone()` which is cheap because that will clone an `imbl::Vector`.